src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+
+			continue
+		}
+
+		runningCaptures++
+
+		log.Info("Add capture waiting group")
+
+		err = api.registerAgentCaptureStream(ctx, &captureStream, log)
+		if err != nil {
+			//TODO: implement error handling
+		}
+
+		c := readMsgFromStream(ctx, captureStream, target, api.bufConf.Size)
+		captureCs = append(captureCs, c)
+	}
+
+	if runningCaptures == 0 {
+		log.Error("Starting of all captures failed during stream preparation")
+		return nil, errorf(codes.FailedPrecondition, "Starting of all captures failed")
+	}
+
+	// merge channels to one channel and send to forward to stream
+	out := mergeResponseChannels(captureCs, api.bufConf.Size)
+	return out, nil
+}
+
+// starts the API drain process, where all ongoing captures are sent a Stop signal so they can terminate gracefully.
+// Runs with a drain timeout and will return an error if the context times out.
+func (api *API) Drain() error {
+	zap.L().Debug("Starting drain with timeout", zap.Duration("timeout", api.drainTimeout))
+	drainTimeout, cancel := context.WithTimeout(context.Background(), api.drainTimeout)
+	defer cancel()
+
+	drained := api.drainStreams()
+	for {
+		select {
+		case <-drained:
+			zap.L().Info("Drain completed successfully.")
+			return nil
+		case <-drainTimeout.Done():
+			zap.L().Warn("Drain timeout reached, but not all clients completed.")
+			return drainTimeout.Err()
+		}
+	}
+}
+
+func (api *API) drainStreams() chan struct{} {
+	defer api.captureLock.Unlock()
+	api.captureLock.Lock()
+
+	api.draining = true
+
+	wg := &sync.WaitGroup{}
+	for client, clientStreams := range api.captures {
+		for vcapID, _ := range clientStreams {
+			zap.L().Debug("Terminating capture for client", zap.String("client", client), zap.String(LogKeyVcapID, vcapID))
+
+			err := api.drainStreamForVcapID(vcapID, wg)
+			if err != nil {
+				zap.S().Warn("Could not send stop request to agent capturing ", zap.String(LogKeyVcapID, vcapID), zap.Error(err))
+				continue
+			}
+		}
+	}
+
+	drained := make(chan struct{})
+
+	go func(wg *sync.WaitGroup) {
+		wg.Wait()
+		close(drained)
+	}(wg)
+
+	return drained
+}
+
+func (api *API) drainStreamForVcapID(vcapID string, wg *sync.WaitGroup) error {
+	for _, agentStream := range api.agentCapturesPerVcapID[vcapID] {
+		err := (*agentStream).Send(&AgentRequest{
+			Payload: &AgentRequest_Stop{},
+		})
+		if err != nil {
+			zap.S().Warn("Could not send stop request to agent capturing request with vcapId", zap.String(LogKeyVcapID, vcapID), zap.Error(err))
+			continue
+		}
+		// The message could be sent, so we expect the context to finish gracefully.
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, ctx context.Context) {
+			// wait for capture stream context do be done
+			<-ctx.Done()
+			wg.Done()
+		}(wg, (*agentStream).Context())
+	}
+	return nil
+}
+
+func (api *API) registerAgentCaptureStream(ctx context.Context, captureStream *captureReceiver, log *zap.Logger) error {
+	defer api.captureLock.Unlock()
+	api.captureLock.Lock()
```

Comment:
Defer the unlock only after acquiring the lock.

Reply:
refactored